cmd/vicadmin/server.go

```diff
@@ -114,39 +137,203 @@ func (s *server) listenPort() int {
 }
 
+// Enforces authentication on route `link` and runs `handler` on successful auth
+func (s *server) AuthenticatedHandle(link string, h http.Handler) {
+	s.Authenticated(link, h.ServeHTTP)
+}
+
+func (s *server) Handle(link string, h http.Handler) {
+	s.mux.Handle(link, gorillacontext.ClearHandler(h))
+}
+
+// Enforces authentication on route `link` and runs `handler` on successful auth
+func (s *server) Authenticated(link string, handler func(http.ResponseWriter, *http.Request)) {
+	defer trace.End(trace.Begin(""))
+
+	authHandler := func(w http.ResponseWriter, r *http.Request) {
+		websession, _ := s.uss.cookies.Get(r, sessionCookieKey) // ignore error because it is okay if it doesn't exist
+
+		if len(r.TLS.PeerCertificates) > 0 { // the user is authenticated by certificate at connection time
+			usersess := s.uss.Add("root", &rootConfig.Config)
+
+			timeNow, err := usersess.created.MarshalText()
+			if err != nil {
+				// it's probably safe to ignore this error since we just created usersess.created when we called Add() above
+				// but just in case..
+				log.Errorf("Failed to unmarshal time object %+v into text due to error: %s", usersess.created, err)
+				http.Redirect(w, r, authFailure, 302)
+				return
+			}
+
+			websession.Values[sessionCreationTimeKey] = string(timeNow)
+			websession.Values[sessionUsernameKey] = "root"
+			err = websession.Save(r, w)
+			if err != nil {
+				log.Errorf("Could not create session for user authenticated via client certificate due to error \"%s\"", err.Error())
+				http.Redirect(w, r, authFailure, 302)
+				return
+			}
+
+			// user was authenticated via cert
+			handler(w, r)
+			return
+		}
+
+		c := websession.Values[sessionCreationTimeKey]
+		if c == nil { // no cookie, so redirect to login
+			log.Errorf("No authentication token: %+v", websession.Values)
+			http.Redirect(w, r, authFailure, 302)
+			return
+		}
+
+		// here we have a cookie, but we need to make sure it's not expired:
+		// parse the cookie creation time
+		created, err := time.Parse(time.RFC3339, c.(string))
+		if err != nil {
+			// we pulled this value out of a cookie, so if it doesn't parse, it might've been tampered with
+			// though the cookie's encrypted so that would destroy the whole cookie..
+			// Handling the error in any case:
+			log.Errorf("Couldn't parse time out of retrieved cookie due to error %s", err.Error())
+			http.Redirect(w, r, authFailure, 302)
+			return
+		}
+
+		// cookie exists but is expired
+		if time.Since(created) > sessionExpiration {
+			http.Redirect(w, r, loginPagePath+"?expired", 302)
+			return
+		}
+
+		// if the date on the cookie was valid, then the user is authenticated
+		handler(w, r)
+	}
+	s.mux.Handle(link, gorillacontext.ClearHandler(http.HandlerFunc(authHandler)))
+}
+
+// renders the page for login and handles authorization requests
+func (s *server) loginPage(res http.ResponseWriter, req *http.Request) {
+	defer trace.End(trace.Begin(""))
+	ctx := context.Background()
```

Comment:
no, this is a different type of context object
https://golang.org/x/net/context